pytorch_toolkit/formula_recognition/im2latex/data/vocab.py

```diff
@@ -0,0 +1,162 @@
+        else:
+            indices_to_convert = indices
+
+        for token in indices_to_convert:
+            val = token.item()
+            if val == END_TOKEN:
+                break
+            phrase_converted.append(
+                self.id2sign.get(val, "?"))
+
+        return " ".join(phrase_converted)
+
+    def construct_indices(self, phrase):
+        """
+        given a phrase, returns indices which
+        this phrase corresponds to
+        """
+        indices = []
+
+        for token in phrase.split():
+            indices.append(self.sign2id.get(token))
+        return torch.Tensor(indices)
+
+
+def write_vocab(data_dir, json=True):
+    """
+    traverse training formulas to make vocab
+    and store the vocab in the file
+    """
+    vocab = Vocab()
+    formulas_file = join(data_dir, 'formulas.norm.lst')
+    with open(formulas_file, 'r') as f:
+        formulas = [formula.strip('\n') for formula in f.readlines()]
+
+    with open(join(data_dir, 'train_filter.lst'), 'r') as f:
+
+        for line in f:
+            _, idx = line.strip('\n').split()
+            idx = int(idx)
+            formula = formulas[idx].split()
+            formula_splitted_numbers = []
+            for sign in formula:
+                formula_splitted_numbers += split_number(sign)
+
+            vocab.add_formula(formula_splitted_numbers)
+
+    vocab_file = join(data_dir, 'vocab.pkl')
+    print("Writing Vocab File in ", vocab_file)
+    with open(vocab_file, 'wb') as w:
+        dict_to_store = {
+            "id2sign": vocab.id2sign,
+            "sign2id": vocab.sign2id,
+        }
+        if json:
+            json.dump(dict_to_store,
+                      w, indent=4, sort_keys=True)
+        else:
+            pkl.dump(dict_to_store, w)
+
+
+def read_vocab(vocab_path):
+    if '.pkl' in vocab_path:
+        with open(vocab_path, "rb") as f:
+            vocab_dict = pkl.load(f)
+    elif 'json' in vocab_path:
+        with open(vocab_path, "r") as f:
+            vocab_dict = json.load(f)
+            vocab_dict['id2sign'] = {int(k): v for k, v in vocab_dict['id2sign'].items()}
+    else:
+        raise ValueError("Wrong extension of the vocab file")
+    vocab = Vocab(loaded_id2sign=vocab_dict["id2sign"], loaded_sign2id=vocab_dict["sign2id"])
+    return vocab
+
+
+def pkl_to_json(vocab_path):
+    with open(vocab_path, "rb") as f:
+        vocab_dict = pkl.load(f)
+        dict_to_store = {
+            "id2sign": vocab_dict["id2sign"],
+            "sign2id": vocab_dict["sign2id"],
+        }
+        json_path = vocab_path.replace(".pkl", ".json")
+        with open(json_path, 'w') as dest:
+            json.dump(dict_to_store, dest, indent=4, sort_keys=True)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Read file with formulas and create vocab file")
+    parser.add_argument('--data_path', help='path to the formulas')
+    args = parser.parse_args()
+    write_vocab(args.data_path)
```

Comment:
I remember why this code is here.      
But it seems to me that it is better to move it to a separate script.    
IMHO, it is not a good idea to have a callable script not in the `tools/` folder.    
Could you make a separate (and very small) script `tools/make_vocab.py`?

Reply:
Done